Migration 002: Remove WebView entries with flags

vinyldarkscratch opened the pull request:
As per the bulk update proposed by #4609, this PR introduces the next migration script that looks through the data for any WebView entries that contain flags, and if it finds them, removes them from the compatibility data.  This is due to the fact that WebView does not have flags (as mentioned by @jpmedley).

The logic is as follows:
- Any single-object statements with a flag will simply be set to `"version_added": false`
- Any arrays with a statement containing a flag will have that statement removed
- Any arrays affected that have a single item will be turned into a single statement object

vinyldarkscratch commented:
Certainly, yeah!  Gotta make sure a bulk change script behaves as expected, right?  I'll look into writing some tests soon -- how should we organize them?  Perhaps `test-###-xxxxx` or something similar?

ddbeck commented:
I'm not sure about naming them. On the one hand, we have a convention in the repo already of `test-whatever.js`. But if these are going to stay in the migrations folder, it might make sense to go with `###-migrationname.test.js`, so they stick together.

Maybe @Elchi3 has thoughts. Or maybe we could look to a tool like [Jest](https://jestjs.io/) for a convention (and heck, maybe use Jest as a test runner someday even, if we're feeling bold).

Elchi3 commented:
I have no strong feelings about naming. Usually, as things grow after time, we will come up with conventions and how to organize things. Having them together in the migrations folder sounds good for now. We can always move things around later.

vinyldarkscratch commented:
Just added a test script to this PR!

ddbeck commented on scripts/migrations/002-remove-webview-flags.test.js:
```diff
@@ -0,0 +1,85 @@
+/** Determines if the OS is Windows */
+const IS_WINDOWS = platform() === 'win32';
+
+const removeWebViewFlags = (key, value) => {
```
Can this be imported from `002-remove-webview-flags.js` instead of being copied?

ddbeck commented on scripts/migrations/002-remove-webview-flags.js:
```diff
@@ -0,0 +1,98 @@
+    if (value.support.webview_android !== undefined) {
+      if (Array.isArray(value.support.webview_android)) {
+        var result = [];
+        for (var i = 0; i < value.support.webview_android.length; i++) {
```
Is there a reason you use `var` in this function (as compared to `let` elsewhere)?

vinyldarkscratch replied:
Old habits die hard -- I'm more accustomed to using `var` myself from when I first started JavaScript, so I have no other reason.  Switching this over now!

ddbeck commented on scripts/migrations/002-remove-webview-flags.js:
```diff
@@ -0,0 +1,98 @@
+  * @param {Promise<void>} filename
+  */
+const fixWebViewFlags = (filename) => {
+  let actual   = fs.readFileSync(filename, 'utf-8').trim();
```
Admittedly, a small thing, but I like using what meager guardrails JS gives us 😆

```suggestion
  const actual   = fs.readFileSync(filename, 'utf-8').trim();
```

ddbeck commented on scripts/migrations/002-remove-webview-flags.js:
```diff
@@ -0,0 +1,98 @@
+  */
+const fixWebViewFlags = (filename) => {
+  let actual   = fs.readFileSync(filename, 'utf-8').trim();
+  let expected = JSON.stringify(JSON.parse(actual, removeWebViewFlags), null, 2);
```
```suggestion
  const expected = JSON.stringify(JSON.parse(actual, removeWebViewFlags), null, 2);
```

ddbeck commented on scripts/migrations/002-remove-webview-flags.test.js:
```diff
@@ -0,0 +1,85 @@
+  "test": {
+    "__compat": {
+      "support": {
+        "webview_android": {
```
A suggestion here: make another test case (or expand this one with an array) showing a webview entry that would _not_ be modified with the migration?

vinyldarkscratch commented:
Thanks @ddbeck!  I've addressed all review comments.  :wink:  (It might be best to separate the new linter updates of adding the migration tests into the test function into its own PR, but then again it may be better to keep it all together here!)

Elchi3 commented:
(drive-by comment, sorry about having had the time to look into this properly yet)
Does this script (need to) take into account the scenario described in https://github.com/mdn/browser-compat-data/pull/4936? There we bump versions the point when there is no flag needed.

vinyldarkscratch commented:
Not a problem, @Elchi3!  This PR won't be going anywhere.  😉

No, this migration script doesn't account for such a scenario...but it may be a good idea to implement such a check, especially depending on whether this or the consistency linter gets merged and/or executed first.  At the same time, however, maybe we should consider a fix-consistency.js since it's more of a consistency issue?

Elchi3 commented:
Right, it is more of a consistency thing. Let's tackle that outside of this PR.

vinyldarkscratch commented on scripts/migrations/002-remove-webview-flags.test.js:
```diff
@@ -0,0 +1,156 @@
+
+const { removeWebViewFlags } = require('./002-remove-webview-flags.js');
+
+const tests = [
```
Yeah, that makes sense.  Fixed!

vinyldarkscratch commented:
Thanks for the additional review!  Regarding the `test` call, I mostly sought to hook it up to the same call as we have another script file tested there.  Perhaps, though, we should separate those off into their own test script, specifically `test-scripts` or `test-infra`, that is always run by Travis?

vinyldarkscratch commented:
Whoops, this was not meant to be closed.  I must've deleted this branch accidentally during cleanup...

ddbeck commented:
>  Regarding the `test` call, I mostly sought to hook it up to the same call as we have another script file tested there.

Ah, yeah, OK that makes sense. You can leave it as-is.

> Perhaps we should separate _all_ of those off into their own test script, specifically `test-scripts` or `test-infra`, that is always run by Travis?

This is a better idea and one that can and should happen in its own PR.

ddbeck commented:
@Elchi3 do you want to take a look at this or is it OK to merge this and move on to the next phase of this removal?

vinyldarkscratch commented:
Yeah, a linter’s exactly what we need afterwards.  Now that you mention it, you’re right — the tests for migrations would need to be run once, and only once.

ddbeck commented:
OK, here was my thinking on this: I want there to be a public accounting for the results of the tests for migrations. To get that, we need to include, in CI, running those tests. I'm fine with removing those tests from CI after we complete the migration, but I think it's important that there's some public validation of, yes, we've run the tests and committed to not letting those tests fail.

As for preventing future flags, the schema itself can forbid webview from having flags data.

Elchi3 commented:
Thanks Daniel! That sounds like a plan to me. 
So, I think we should merge this, run the migration, then change the tests to be in linting rather than it testing the migration.

vinyldarkscratch commented:
Seems like we're all ready to run the migration, though we haven't chosen a date just yet?

ddbeck commented:
@vinyldarkscratch Yeah, that's right. Suppose I merged this today. Would you be willing to do the migration PR after this week's release (e.g. later on Thursday or on Friday)? (I'd say sooner, but the start of this week is a bit busy.)

vinyldarkscratch commented:
Sweet, sounds good to me!  Yeah, I can totally get that PR up.  :wink:
